src/utils/internalLinks.js

```diff
@@ -0,0 +1,117 @@
+/* @flow strict-local */
+import type { Narrow, User, Stream } from '../types';
+import { topicNarrow, streamNarrow, groupNarrow, specialNarrow } from './narrow';
+import { isUrlOnRealm } from './url';
+import { getUserById } from '../users/userHelpers';
+import { transformToEncodedURI } from './string';
+import { NULL_USER } from '../nullObjects';
+
+export const getPathsFromUrl = (url: string = '', realm: string) => {
+  const paths = url
+    .split(realm)
+    .pop()
+    .split('#narrow/')
+    .pop()
+    .split('/');
+
+  if (paths.length > 0 && paths[paths.length - 1] === '') {
+    // url ends with /
+    paths.splice(-1, 1);
+  }
+  return paths;
+};
+
+export const isInternalLink = (url: string, realm: string): boolean =>
+  isUrlOnRealm(url, realm) ? /^(\/#narrow|#narrow)/i.test(url.split(realm).pop()) : false;
+
+export const isMessageLink = (url: string, realm: string): boolean =>
+  isInternalLink(url, realm) && url.includes('near');
+
+type LinkType = 'external' | 'home' | 'group' | 'topic' | 'stream' | 'special';
+
+export const getLinkType = (url: string, realm: string): LinkType => {
+  if (!isInternalLink(url, realm)) {
+    return 'external';
+  }
+
+  const paths = getPathsFromUrl(url, realm);
+
+  if (
+    (paths.length === 2 && paths[0] === 'pm-with')
+    || (paths.length === 4 && paths[0] === 'pm-with' && paths[2] === 'near')
+  ) {
+    return 'group';
+  }
+
+  if (
+    (paths.length === 4 || paths.length === 6)
+    && paths[0] === 'stream'
+    && (paths[2] === 'subject' || paths[2] === 'topic')
+  ) {
+    return 'topic';
+  }
+
+  if (paths.length === 2 && paths[0] === 'stream') {
+    return 'stream';
+  }
+
+  if (paths.length === 2 && paths[0] === 'is' && /^(private|starred|mentioned)/i.test(paths[1])) {
+    return 'special';
+  }
+
+  return 'home';
+};
+
+export const getNarrowFromLink = (
+  url: string,
+  realm: string,
+  users: User[],
+  streams: Stream[],
+): ?Narrow => {
+  const type = getLinkType(url, realm);
```

Comment:
The commit that switches to using this is best squashed with its parent which creates this helper. That makes it a lot clearer why the details of the helper are as they are -- they're just a factoring-out of existing logic in this function.

If that helper `getLinkType` were being introduced fresh, I would have a lot of questions to address before wanting to merge it. ;-) See comment above for some of them. But when it's just moving logic we already have and making it better, it gets a pass on a lot of that.

Reply:
Squashed.